var/spack/repos/builtin/packages/glibc/package.py

```diff
@@ -0,0 +1,76 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+import os
+
+
+class Glibc(Package):
+    """Any Unix-like operating system needs a C library: the library which
+    defines the ``system calls'' and other basic facilities such as open,
+    malloc, printf, exit... The GNU C Library is used as the C library in the
+    GNU system and in GNU/Linux systems, as well as many other systems that use
+    Linux as the kernel."""
+
+    homepage = "https://www.gnu.org/software/libc/"
+    url = "http://open-source-box.org/glibc/glibc-2.23.tar.xz"
+
+    version('2.23', '456995968f3acadbed39f5eba31678df')
+    version('2.22', 'e51e02bf552a0a1fbbdc948fb2f5e83c')
+    version('2.21', '9cb398828e8f84f57d1f7d5588cf40cd')
+    version('2.20', '948a6e06419a01bd51e97206861595b0')
+    version('2.19', 'e26b8cc666b162f999404b03970f14e4')
+    version('2.18', '88fbbceafee809e82efd52efa1e3c58f')
+    version('2.17', '87bf675c8ee523ebda4803e8e1cec638')
+    version('2.16.0', '80b181b02ab249524ec92822c0174cf7')
+
+    provides('libc')
+
+    depends_on('binutils@2.20:')
+
+    # Linux kernel headers 2.6.19 or later are required
+    # depends_on("linux-headers")
+
+    dynamic_linker = 'lib/ld-linux-x86-64.so.2'
```

Comment:
You can't access `self` outside instance methods, but assuming that the path is prefix-relative is fine.

@adamjstewart: Linking with `libc` is a bit more complicated than other dependencies, as you need to specify a loader when you link.  You can't use the default `ld.so`, as it is tied to the system `glibc` on Linux.  Also, building/linking with a custom glibc will likely break ABI compatibility with the system `glibc`, so everything in a DAG needs to depend on the same `glibc`.  You *can* have different `glibc`'s in separate DAGs.

We need to be careful this.  Technically, nearly everything depends on with `libc` implicitly, but that is usually through the compiler and I would rather not expose package authors to this complexity.  Ideally, only packages that really *need* a custom `glibc` will depend on it for now, until I can make our compiler dependency model more robust.  I'm trying to figure out a way to get @mwilliammyers's support for glibc into spack in a way that doesn't disrupt other stuff.